GradeServer.c

```diff
@@ -1,146 +1,216 @@
-     printf("ID: %s, Password: %s\n", assistants[i].id, assistants[i].password);
-  }
-
-  // Init task list and locks
-  task_list_size = 0;
-  pthread_mutex_init(&task_list_mutex, NULL);
-  pthread_cond_init(&task_list_cond, NULL);
-
-  // Get the port number from the command line argument
-  int port = atoi(argv[1]);
-
-  // Create a socket
-  sockfd = socket(AF_INET, SOCK_STREAM, 0);
-  if (sockfd < 0) {
-    perror("Error opening socket");
-    return 1;
-  }
-
-  // Set up the server address
-  struct sockaddr_in serv_addr;
-  memset(&serv_addr, 0, sizeof(serv_addr));
-  serv_addr.sin_family = AF_INET;
-  serv_addr.sin_addr.s_addr = INADDR_ANY;
-  serv_addr.sin_port = htons(port);
-
-  // Bind the socket to the server address
-  if (bind(sockfd, (struct sockaddr*) &serv_addr, sizeof(serv_addr)) < 0) {
-    perror("Error binding socket");
-    return 1;
-  }
-
-  // Listen for incoming connections
-  listen(sockfd, 5);
-
-  // Create a pool of worker threads
-  pthread_t threads[NUM_THREADS];
-  for (int i = 0; i < NUM_THREADS; i++) {
-    if (pthread_create(&threads[i], NULL, handle_connection, NULL) != 0) {
-      perror("Error creating thread");
-      return 1;
-    }
-  }
-
-  // Infinite loop - accept user's connection and adding task for every new connection
-  while (1) {
-    // Accept a connection
-    struct sockaddr_in cli_addr;
-    socklen_t cli_addr_len = sizeof(cli_addr);
-    int connfd = accept(sockfd, (struct sockaddr*) &cli_addr, &cli_addr_len);
-    if (connfd < 0) {
-      perror("Error accepting connection");
-      return 1;
+void delete_task(task_t task)
+{
+    pthread_mutex_lock(&task_mutex);
+
+    for (int i = 0; i < num_tasks; i++)
+    {
+        if (strcmp(tasks[i].message, task.message) == 0)
+        {
+            // Shift all tasks after the specified index one position to the left
+            for (int j = i; j < num_tasks - 1; j++)
+                tasks[j] = tasks[j + 1];
+
+            num_tasks--;
+            break;
+        }
     }
-      add_task_to_queue(conndf);
-  }
-  // TODO - this part never gets called because of the while loop before!
-  // Wait for the worker threads to finish
-  for (int i = 0; i < NUM_THREADS; i++) {
-    pthread_join(threads[i], NULL);
-  }
-
-  // Close the socket
-  close(sockfd);
-
-  // Free memory and destroy synchronization variables
-  free(task_list);
-  pthread_mutex_destroy(&task_list_mutex);
-  pthread_cond_destroy(&task_list_cond);
-
-  return 0;
+
+    pthread_mutex_unlock(&task_mutex);
 }
 
+void process_task(task_t task)
+{
+    // Do something with the task
+    char* command[2000];
+    strcpy(command, task.message);
+
+    switch (command) {
+        // TODO - should we validate no additional arguments?
 
-void sigint_handler(int sig) {
-  printf("Received SIGINT, shutting down...\n");
+        case "Login":
+            login(arg1, arg2, response);
+            break;
 
-  // Close the socket and exit
-  close(sockfd);
```

Comment:
In the old version we'd close the socket but you removed it - we don't need it?

Reply:
Yeah I guess that we need to close this socket...